model/scope.go

```diff
@@ -0,0 +1,168 @@
+package model
+
+import (
+	"encoding/json"
+	"strings"
+)
+
+type Scope []string
+
+const (
+	ScopeChannelJoin  = "channels:join"
+	ScopeMessageRead  = "message:read"
+	ScopeMessageWrite = "message:write"
+	ScopeFilesRead    = "files:read"
+	ScopeFilesWrite   = "files:write"
+	ScopeSearchPosts  = "search:post"
+)
+
+func ScopeAny(scopes ...string) Scope {
+	return scopes
+}
+
+func ScopeAllow() Scope {
+	return Scope{}
+}
+
+func ScopeDeny() Scope {
+	return nil
+}
+
+func (s Scope) IsInScope(scope string) bool {
+	if s == nil {
+		return false
+	}
+
+	for _, allowed := range s {
+		if scope == allowed {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (s Scope) IsPluginInScope(pluginID string) bool {
+	if s == nil {
+		return false
+	}
+
+	for _, allowed := range s {
+		if strings.HasPrefix(allowed, "plugin:") &&
```

Comment:
`HasPrefix` is redundant here. If prefix isn't there, the `TrimPrefix` comparison would anyways fail.

Reply:
Not necessarily. Imagine the scope "message:read". It doesn't have the prefix, and trimPrefix will return the string untouched, like "message:read". Now imagine that for whatever reason, someone decided to have its plugin id as "message:read". Now the plugin would be considered allowed.

It is uncommon and not very exploitable? Yes. But still, I prefer to cover that case.